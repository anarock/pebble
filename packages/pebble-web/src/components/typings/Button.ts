import * as React from "react";
import * as CSS from "csstype";
import { Interpolation } from "./emotionCustom";

export type ButtonType = "primary" | "secondary" | "link" | "success" | "alert";

export interface ButtonProps {
  children: React.ReactNode;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
  showShadow?: boolean;
  disabled?: boolean;
  type?: ButtonType;
  size?: "x-small" | "small" | "large";
  width?: CSS.WidthProperty<number>;
  styles?: Interpolation;
  showRipple?: boolean;
  loading?: boolean;
  filled?: boolean;
  buttonProps?: React.ButtonHTMLAttributes<HTMLButtonElement>;
}

export interface DropDownButtonProps extends ButtonProps {
  isOpen: boolean;
  isSelected: boolean;
}

interface MappingColorByTypeOptions {
  textColor?: string;
  base: string;
  hover: string;
  active: string;
  disabled: string;
}

export interface MappingColorByType {
  primary: MappingColorByTypeOptions;
  secondary: MappingColorByTypeOptions;
  success: MappingColorByTypeOptions;
  alert: MappingColorByTypeOptions;
}
