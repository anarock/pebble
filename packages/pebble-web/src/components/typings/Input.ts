import * as React from "react";
import { Interpolation } from "./emotionCustom";

interface CommonInputProps {
  placeholder: string;
  onChange: (text: string) => void;
  value?: string | number;
  styles?: Interpolation;
  disabled?: boolean;
  errorMessage?: string;
  fixLabelAtTop?: boolean;
  inputStyles?: Interpolation;
  loading?: boolean;
  message?: string;
  onClick?: React.MouseEventHandler<HTMLDivElement>;
  readOnly?: boolean;
  required?: boolean;
  successMessage?: string;
}

export interface SimpleInputProps extends CommonInputProps {
  inputProps?: React.InputHTMLAttributes<HTMLInputElement> &
    React.RefAttributes<HTMLInputElement>;
  textArea?: false;
  type?: "text" | "date" | "password" | "number" | "email" | "tel";
}

interface TextAreaInputProps extends CommonInputProps {
  inputProps?: React.InputHTMLAttributes<HTMLTextAreaElement>;
  textArea: true;
  type?: undefined;
}

export type InputProps = SimpleInputProps | TextAreaInputProps;

export interface InputState {
  isFocused: boolean;
}
