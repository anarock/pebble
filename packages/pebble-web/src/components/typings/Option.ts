import { ControlProps } from "./Control";
import { Omit } from "utility-types";
import * as React from "react";
import { Interpolation } from "./emotionCustom";

export interface OptionProps<OptionType>
  extends Omit<ControlProps<OptionType>, "checked" | "type"> {
  isActive?: boolean;
  isSelected?: boolean;
  multiSelect?: boolean;
  labelStyles?: Interpolation;
  rightElement?: (props: OptionProps<OptionType>) => React.ReactNode;
}
