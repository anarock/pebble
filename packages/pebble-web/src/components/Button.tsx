import * as React from "react";

import {
  iconStyle,
  getButtonStyle,
  dropDownButtonStyle,
  dropDownButtonDefaultStyle
} from "./styles/Button.styles";
import Ink from "react-ink";
import { ButtonProps, DropDownButtonProps } from "./typings/Button";
import Loader from "./Loader";
import { colors } from "pebble-shared";

const Button: React.FunctionComponent<ButtonProps> = ({
  type = "primary",
  disabled,
  children,
  onClick,
  width,
  showShadow,
  styles,
  showRipple = true,
  loading,
  size = "small",
  buttonProps
}: ButtonProps) => {
  const disableAction = disabled || loading;

  const filled = size !== "x-small";

  return (
    <button
      css={[getButtonStyle(size, type, !!showShadow, filled), styles]}
      onClick={!disableAction ? onClick : undefined}
      style={{ width }}
      disabled={disabled}
      {...buttonProps}
    >
      {loading ? <Loader color={colors.white.base} scale={0.4} /> : children}
      {!disableAction && showRipple && type !== "link" && <Ink />}
    </button>
  );
};

export const DropDownButton = ({
  isOpen,
  isSelected,
  children,
  styles,
  ...props
}: DropDownButtonProps) => {
  return (
    <Button
      {...props}
      type="secondary"
      styles={[
        dropDownButtonStyle,
        !(isOpen || isSelected) && dropDownButtonDefaultStyle,
        styles
      ]}
    >
      {children}{" "}
      <i
        className="pi pi-arrow-drop-down"
        css={iconStyle}
        style={{
          transform: isOpen ? "rotate(180deg)" : "none"
        }}
      />
    </Button>
  );
};

export default Button;
