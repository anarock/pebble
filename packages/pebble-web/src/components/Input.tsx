import * as React from "react";

import { InputProps, InputState } from "./typings/Input";
import {
  highlightStyle,
  labelStyle,
  wrapperStyle,
  messageStyle,
  inputStyle,
  inputReadOnlyStyle,
  inputDisabledStyle,
  inputTextAreaStyle,
  loadingStyle
} from "./styles/Input.styles";
import { colors } from "pebble-shared";
import Loader from "./Loader";

function getColor(
  error: string | undefined,
  success: string | undefined,
  isUnderlineColor?: boolean
) {
  let color = colors.gray.dark;
  if (error) {
    color = colors.red.base;
  } else if (success) {
    color = colors.emerald.base;
  } else if (isUnderlineColor) {
    color = colors.violet.base;
  }

  return color;
}

class Input extends React.PureComponent<InputProps, InputState> {
  readonly state: InputState = {
    isFocused: false
  };

  static defaultProps = {
    value: "",
    readOnly: false
  };

  private addFocus = () => {
    this.setState({
      isFocused: true
    });
  };

  private removeFocus = () => {
    this.setState({
      isFocused: false
    });
  };

  private handleChange = (
    e:
      | React.ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    this.props.onChange(e.target.value || "");
  };

  render() {
    const {
      type,
      placeholder,
      styles,
      inputStyles: overridenInputStyles,
      fixLabelAtTop,
      value,
      readOnly,
      disabled,
      errorMessage,
      successMessage,
      message,
      textArea,
      required,
      onClick,
      loading
    } = this.props;
    const { isFocused } = this.state;

    const _message = errorMessage || successMessage || message;

    const _inputClassName = [
      inputStyle,
      !!disabled && inputDisabledStyle,
      !!readOnly && inputReadOnlyStyle,
      !!textArea && inputTextAreaStyle,
      overridenInputStyles
    ];

    const _inputProps = {
      "aria-label": placeholder ? placeholder : undefined,
      css: _inputClassName,
      disabled,
      onChange: this.handleChange,
      readOnly,
      value: value || ""
    };

    const highlightClassName = [
      !!isFocused && "_pebble_input_highlight_focused",
      !!errorMessage || (!!successMessage && "_pebble_input_highlight_state"),
      !!readOnly && "_pebble_input_highlight_read_only",
      !!disabled && "_pebble_input_highlight_disabled"
    ]
      .filter(f => f)
      .join(" ");

    const labelClassName = !!(isFocused || !!value || fixLabelAtTop)
      ? "_pebble_input_label_focused"
      : "";

    return (
      <div
        css={[wrapperStyle, styles]}
        className={textArea ? "_pebble_input_wrapper_textarea" : ""}
        onFocus={this.addFocus}
        onBlur={this.removeFocus}
        onClick={onClick}
      >
        {this.props.textArea ? (
          <textarea {..._inputProps} {...this.props.inputProps} />
        ) : (
          <input type={type} {..._inputProps} {...this.props.inputProps} />
        )}

        <label css={labelStyle} className={labelClassName}>
          {placeholder}
          {required && <span style={{ color: colors.red.base }}>&nbsp;*</span>}
        </label>

        <div
          css={highlightStyle}
          className={highlightClassName}
          style={{
            backgroundColor: getColor(errorMessage, successMessage, true)
          }}
        />

        {loading && (
          <Loader
            color={colors.violet.base}
            scale={0.6}
            styles={loadingStyle}
          />
        )}

        {_message && (
          <div
            css={messageStyle}
            style={{ color: getColor(errorMessage, successMessage) }}
          >
            {_message}
          </div>
        )}
      </div>
    );
  }
}

export default Input;
