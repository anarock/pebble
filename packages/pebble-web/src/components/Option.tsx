import * as React from "react";
import { OptionProps } from "./typings/Option";
import Control from "./shared/Control";
import {
  activeRow,
  rowWrapper,
  selectedRow,
  labelWrap
} from "./styles/Options.styles";
import Ink from "react-ink";
import { colors } from "pebble-shared";

const defaultProps = {
  rightElement: ({
    isSelected,
    multiSelect
  }: {
    isSelected: boolean;
    multiSelect: boolean;
  }) => {
    const iconClass = [
      "pi",
      isSelected && "pi-checkbox-selected",
      !isSelected && "pi-checkbox-unselected"
    ]
      .filter(Boolean)
      .join(" ");
    return multiSelect ? (
      <i
        className={iconClass}
        css={{
          marginLeft: "10px",
          color: isSelected ? colors.violet.base : colors.gray.light,
          fontSize: "20px"
        }}
      />
    ) : null;
  }
};

class Option<OptionType> extends React.Component<
  OptionProps<OptionType> & Required<typeof defaultProps>
> {
  static defaultProps = defaultProps;
  render() {
    const {
      label,
      isActive,
      isSelected,
      rightElement,
      labelStyles,
      styles
    } = this.props;
    const _class = [
      rowWrapper,
      !!isActive && activeRow,
      !!isSelected && selectedRow,
      styles
    ];
    return (
      <Control
        {...this.props}
        checked={this.props.isSelected}
        type={this.props.multiSelect ? "checkbox" : "radio"}
        styles={_class}
      >
        {() => {
          return (
            <>
              <div css={[labelWrap, labelStyles]}>{label}</div>
              {rightElement(this.props)}
              <Ink />
            </>
          );
        }}
      </Control>
    );
  }
}

export default Option;
