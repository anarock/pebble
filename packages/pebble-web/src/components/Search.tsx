import * as React from "react";

import { SearchProps } from "./typings/Search";
import {
  searchStyle,
  searchWrapperStyle,
  clearContainer
} from "./styles/Search.styles";

class Search extends React.PureComponent<SearchProps> {
  searchInputRef: React.RefObject<HTMLInputElement> = React.createRef();

  static defaultProps = {
    showSearchIcon: true,
    clearable: true
  };

  render() {
    const {
      type,
      inputProps,
      onChange,
      placeholder,
      showSearchIcon,
      styles,
      clearable,
      value
    } = this.props;

    const wrapperClassName = [
      type === "small" && "__pebble__search__small",
      type === "large" && "__pebble__search__large",
      type === "table" && "__pebble__search__table"
    ]
      .filter(Boolean)
      .join(" ");

    return (
      <div className={wrapperClassName} css={[searchWrapperStyle, styles]}>
        {type !== "large" && showSearchIcon && <i className="pi pi-search" />}
        <input
          css={searchStyle}
          type="text"
          aria-label={placeholder}
          placeholder={placeholder}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
            onChange(e.target.value);
          }}
          ref={this.searchInputRef}
          value={value}
          {...inputProps}
        />
        {clearable && (
          <div
            className={(!!value && !!value.length && "__display") || ""}
            css={clearContainer}
            onClick={() => {
              if (this.searchInputRef.current) {
                this.searchInputRef.current.value = "";
              }
              onChange("");
            }}
          >
            <i
              className="pi pi-close"
              style={{ display: "table-cell", verticalAlign: "middle" }}
            />
          </div>
        )}
      </div>
    );
  }
}

export default Search;
