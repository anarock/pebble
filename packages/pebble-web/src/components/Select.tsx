import * as React from "react";

import { SelectProps } from "./typings/Select";
import {
  chevronStyle,
  dropDownClass,
  inputWrapper,
  selectInput,
  selectInputWrapper,
  selectWrapper,
  fullWidth,
  relativePosition
} from "./styles/Select.styles";
import DropDown from "./DropDown";
import Input from "./Input";
import OptionGroupCheckBox from "./OptionGroupCheckBox";
import OptionGroupRadio from "./OptionGroupRadio";

function noop() {}

function Select<OptionType>(props: SelectProps<OptionType>) {
  const {
    styles,
    placeholder,
    required,
    errorMessage,
    value,
    dropdownStyles,
    inputProps,
    fullWidthDropdown,
    onDropdownToggle = noop,
    disabled,
    isSelected,
    placement,
    modifiers
  } = props;

  return (
    <div css={[selectWrapper, styles, !!fullWidthDropdown && relativePosition]}>
      <DropDown
        dropDownStyles={[
          dropDownClass,
          dropdownStyles,
          !!fullWidthDropdown && fullWidth
        ]}
        onOutsideClick={isOpen => onDropdownToggle(isOpen)}
        labelComponent={({ toggleDropdown, isOpen }) => {
          const chevron = [
            "pi",
            "pi-arrow-drop-down",
            isOpen && "__pebble__select__open"
          ]
            .filter(Boolean)
            .join(" ");
          return (
            <div
              css={inputWrapper}
              onClick={
                disabled
                  ? undefined
                  : () => {
                      toggleDropdown();
                      onDropdownToggle(isOpen);
                    }
              }
            >
              <Input
                styles={selectInputWrapper}
                inputStyles={selectInput}
                placeholder={placeholder}
                value={value || ""}
                onChange={noop}
                required={required}
                message={isOpen ? " " : ""}
                errorMessage={errorMessage}
                readOnly
                disabled={disabled}
                {...inputProps}
              />
              <i css={chevronStyle} className={chevron} />
            </div>
          );
        }}
        placement={placement}
        modifiers={modifiers}
      >
        {({ toggle, isOpen }) => {
          const { children, onClear, searchBox, searchBoxProps } = props;
          const commonProps = {
            isSelected,
            onClear:
              onClear &&
              (() => {
                onClear();
                onDropdownToggle(isOpen);
                toggle();
              }),
            searchBox,
            searchBoxProps
          };

          // This would have been the ideal way to write this but typescript is crying.
          // const OptionGroup = props.multiSelect
          //   ? OptionGroupCheckBox
          //   : OptionGroupRadio;
          // return (
          //   <OptionGroup
          //     selected={props.selected}
          //     onChange={(_value, extras) => {
          //       if (_value) {
          //         props.onChange(_value, extras);
          //       }
          //       if (!props.multiSelect) {
          //         onDropdownToggle(isOpen);
          //         toggle();
          //       }
          //     }}
          //     onApply={
          //       props.multiSelect &&
          //       props.onApply &&
          //       (_value => {
          //         if (props.onApply) {
          //           props.onApply(_value, props);
          //         }
          //         onDropdownToggle(isOpen);
          //         toggle();
          //       })
          //     }
          //     {...commonProps}
          //   >
          //     {children}
          //   </OptionGroup>
          // );

          if (props.multiSelect) {
            return (
              <OptionGroupCheckBox<OptionType>
                selected={props.selected}
                onChange={(_value, extras) => {
                  props.onChange(_value, extras);
                }}
                onApply={
                  props.onApply &&
                  (_value => {
                    if (props.onApply) props.onApply(_value, props);
                    onDropdownToggle(isOpen);
                    toggle();
                  })
                }
                {...commonProps}
              >
                {children}
              </OptionGroupCheckBox>
            );
          } else {
            return (
              <OptionGroupRadio
                selected={props.selected}
                onChange={(_value, extras) => {
                  if (_value) props.onChange(_value, extras);
                  onDropdownToggle(isOpen);
                  toggle();
                }}
                {...commonProps}
              >
                {children}
              </OptionGroupRadio>
            );
          }
        }}
      </DropDown>
    </div>
  );
}

export default Select;
