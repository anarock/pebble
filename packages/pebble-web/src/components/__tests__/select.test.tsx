import * as React from "react";
import renderer from "react-test-renderer";
import Option from "../Option";
import Select from "../Select";
import { mount } from "enzyme";
import sinon from "sinon";
import { SingleSelectProps, MultiSelectProps } from "../typings/Select";
import Button from "../Button";
import Input from "../Input";
import Search from "../Search";
import "../../../tests/__setup__/matchers";

const options = new Array(5)
  .fill(1)
  .map((_x, i) => (
    <Option key={i + 1} value={`option-${i + 1}`} label="I am an option" />
  ));

const noop = () => {};

function getComponent(
  spy = noop,
  props: Partial<SingleSelectProps<string>> = {}
) {
  return (
    <Select
      onChange={spy}
      placeholder="Choose Option"
      selected={"option-2"}
      {...props}
    >
      {options}
    </Select>
  );
}

function getMultiSelectComponent<T>(
  spy = noop,
  props: Partial<MultiSelectProps<T>> = {}
) {
  return (
    <Select<T>
      onChange={spy}
      placeholder="Choose Option"
      multiSelect
      {...props}
    >
      {options}
    </Select>
  );
}

describe("Component: Select", () => {
  test("single-select: snapshot", () => {
    const select = renderer.create(getComponent());
    const tree = select.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("multi-select: snapshot", () => {
    const select = renderer.create(
      getMultiSelectComponent(undefined, {
        multiSelect: true,
        selected: []
      })
    );
    const tree = select.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("single-select with searchbox: snapshot", () => {
    const select = renderer.create(
      getComponent(undefined, {
        searchBox: true,
        searchBoxProps: {
          placeholder: "Search",
          onChange: () => {},
          value: ""
        }
      })
    );
    const tree = select.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("should open when input is clicked", () => {
    const spy = sinon.spy();
    const select = mount(getComponent(spy));

    select.find(Input).simulate("click");
    expect(select.find(Option)).toHaveLength(5);
  });

  test("single select: should trigger onChange with correct onChange", () => {
    const clock = sinon.useFakeTimers();
    const spy = sinon.spy();
    const select = mount(getComponent(spy));
    select.find(Input).simulate("click");
    select
      .find(Option)
      .at(2)
      .simulate("click");

    expect(spy.calledWith("option-3")).toBeTruthy();

    // wait for the dropdown animation to get over.
    clock.tick(1000);

    // This means that Option is no more rendered in DOM.
    expect(select).toNotBeInDOM("[data-test-id='optiongroup']");
  });

  test("multi select: should trigger onChange with correct onChange", () => {
    const clock = sinon.useFakeTimers();
    const spy = sinon.spy();
    const applySpy = sinon.spy();
    const clearSpy = sinon.spy();
    const select = mount(
      getMultiSelectComponent(spy, {
        multiSelect: true,
        onApply: applySpy,
        onClear: clearSpy,
        selected: []
      })
    );
    select.find(Input).simulate("click");
    select
      .find(Option)
      .at(2)
      .simulate("click");

    expect(spy.calledWith(["option-3"])).toBeTruthy();

    select.setProps({
      selected: ["option-3"]
    });

    select
      .find(Option)
      .at(3)
      .simulate("click");

    expect(spy.calledWith(["option-3", "option-4"])).toBeTruthy();

    select.setProps({
      selected: ["option-3", "option-4"]
    });

    select
      .find(Option)
      .at(2)
      .simulate("click");

    expect(spy.calledWith(["option-4"])).toBeTruthy();

    expect(select.find(Button)).toHaveLength(2);

    // test onClear
    select
      .find(Button)
      .at(0)
      .simulate("click");
    expect(clearSpy.calledOnce).toBeTruthy();

    // Reopen dropdown and test onApply
    select.find(Input).simulate("click");
    select.setProps({
      selected: ["option-3"]
    });

    select
      .find(Button)
      .at(1)
      .simulate("click");
    expect(applySpy.calledWith(["option-3"])).toBeTruthy();

    clock.tick(1000);

    // ensure the dropdown is closed
    expect(select).toNotBeInDOM("[data-test-id='optiongroup']");
  });

  test("single select: query change triggers onChange", () => {
    const spy = sinon.spy();
    const queryChangeSpy = sinon.spy();

    const select = mount(
      getComponent(spy, {
        searchBox: true,
        searchBoxProps: {
          placeholder: "Search",
          onChange: queryChangeSpy,
          value: ""
        }
      })
    );
    select.find(Input).simulate("click");
    select
      .find(Search)
      .find("input")
      .simulate("change", {
        target: {
          value: "hello"
        }
      });

    expect(queryChangeSpy.calledWith("hello")).toBeTruthy();
  });

  test("single select: clearing query triggers onChange", () => {
    const spy = sinon.spy();
    const clearQuerySpy = sinon.spy();
    const select = mount(
      getComponent(spy, {
        searchBox: true,
        searchBoxProps: {
          placeholder: "Search",
          onChange: clearQuerySpy,
          value: "hello"
        }
      })
    );
    select.find(Input).simulate("click");

    select
      .find(Search)
      .find(".pi-close")
      .hostNodes()
      .simulate("click");
    expect(clearQuerySpy.calledWith("")).toBeTruthy();
  });

  test("single select: onDropdownToggle is triggered correctly", () => {
    const spy = sinon.spy();
    const toggleSpy = sinon.spy();

    const select = mount(
      getComponent(spy, {
        searchBox: true,
        onDropdownToggle: toggleSpy
      })
    );

    // The animation hasn't run yet so the component won't be even present in the vDOM
    expect(select.find(Option)).toHaveLength(0);

    select.find(Input).simulate("click");
    expect(select.find(Option)).toHaveLength(5);
    expect(toggleSpy.calledWith(false)).toBeTruthy();

    select.find(Input).simulate("click");
    expect(toggleSpy.calledWith(true)).toBeTruthy();
  });
});

test("should not open if disabled when input is clicked", () => {
  const spy = sinon.spy();
  const toggleSpy = sinon.spy();

  const select = mount(
    getComponent(spy, {
      searchBox: true,
      onDropdownToggle: toggleSpy
    })
  );

  // The animation hasn't run yet so the component won't be even present in the vDOM
  expect(select.find(Option)).toHaveLength(0);

  select.find(Input).simulate("click");
  expect(toggleSpy.calledWith(false)).toBeTruthy();
});
