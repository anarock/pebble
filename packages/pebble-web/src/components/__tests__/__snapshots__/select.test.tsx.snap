// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Select multi-select: snapshot 1`] = `
.emotion-7 {
  margin-bottom: 20px;
}

.emotion-6 {
  position: relative;
}

.emotion-5 {
  cursor: pointer;
  position: relative;
}

.emotion-3 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #ffffff;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 20px;
  height: 68px;
  pointer-events: none;
  margin-bottom: 0;
}

.emotion-3._pebble_input_wrapper_textarea {
  height: 110px;
}

.emotion-0 {
  outline: 0;
  border: 0;
  border-bottom: 1px solid #EDEDED;
  padding: 24px 0 12px 0;
  height: 48px;
  border-radius: 0;
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  width: 100%;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: #6B7785;
  color: #101721;
  padding-right: 15px;
}

.emotion-0:disabled {
  background-color: #ffffff;
}

.emotion-0::-webkit-input-placeholder {
  color: #E0E0E0;
}

.emotion-0::-moz-placeholder {
  color: #E0E0E0;
}

.emotion-0:-ms-input-placeholder {
  color: #E0E0E0;
}

.emotion-0:-moz-placeholder {
  color: #E0E0E0;
}

.emotion-0::-webkit-input-placeholder {
  color: #E0E0E0;
}

.emotion-0::-moz-placeholder {
  color: #E0E0E0;
}

.emotion-0:-ms-input-placeholder {
  color: #E0E0E0;
}

.emotion-0::placeholder {
  color: #E0E0E0;
}

.emotion-1 {
  color: #9FA6B0;
  font-size: 14px;
  line-height: 12px;
  position: absolute;
  -webkit-transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  -webkit-transform: translate(0,24px);
  -ms-transform: translate(0,24px);
  transform: translate(0,24px);
  pointer-events: none;
}

.emotion-1._pebble_input_label_focused {
  font-size: 12px;
  line-height: 10px;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
  color: #6B7785;
}

.emotion-2 {
  height: 2px;
  background-color: #6161FF;
  margin-top: -2px;
  position: relative;
  width: 0;
  -webkit-transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  z-index: 9;
}

.emotion-2._pebble_input_highlight_state,
.emotion-2._pebble_input_highlight_focused {
  width: 100%;
}

.emotion-4 {
  position: absolute;
  top: 0;
  bottom: 0;
  margin: auto;
  height: 10px;
  right: 7px;
  color: #9FA6B0;
  font-size: 10px;
}

.emotion-4.__pebble__select__open {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

<div
  className="emotion-7"
>
  <div
    className="emotion-6"
  >
    <div
      style={
        Object {
          "display": "inline-block",
          "width": "100%",
        }
      }
    >
      <div
        className="emotion-5"
        onClick={[Function]}
      >
        <div
          className=" emotion-3"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <input
            aria-label="Choose Option"
            className="emotion-0"
            onChange={[Function]}
            readOnly={true}
            value=""
          />
          <label
            className=" emotion-1"
          >
            Choose Option
          </label>
          <div
            className="_pebble_input_highlight_read_only emotion-2"
            style={
              Object {
                "backgroundColor": "#6161FF",
              }
            }
          />
          
        </div>
        <i
          className="pi pi-arrow-drop-down emotion-4"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Component: Select single-select with searchbox: snapshot 1`] = `
.emotion-7 {
  margin-bottom: 20px;
}

.emotion-6 {
  position: relative;
}

.emotion-5 {
  cursor: pointer;
  position: relative;
}

.emotion-3 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #ffffff;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 20px;
  height: 68px;
  pointer-events: none;
  margin-bottom: 0;
}

.emotion-3._pebble_input_wrapper_textarea {
  height: 110px;
}

.emotion-0 {
  outline: 0;
  border: 0;
  border-bottom: 1px solid #EDEDED;
  padding: 24px 0 12px 0;
  height: 48px;
  border-radius: 0;
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  width: 100%;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: #6B7785;
  color: #101721;
  padding-right: 15px;
}

.emotion-0:disabled {
  background-color: #ffffff;
}

.emotion-0::-webkit-input-placeholder {
  color: #E0E0E0;
}

.emotion-0::-moz-placeholder {
  color: #E0E0E0;
}

.emotion-0:-ms-input-placeholder {
  color: #E0E0E0;
}

.emotion-0:-moz-placeholder {
  color: #E0E0E0;
}

.emotion-0::-webkit-input-placeholder {
  color: #E0E0E0;
}

.emotion-0::-moz-placeholder {
  color: #E0E0E0;
}

.emotion-0:-ms-input-placeholder {
  color: #E0E0E0;
}

.emotion-0::placeholder {
  color: #E0E0E0;
}

.emotion-1 {
  color: #9FA6B0;
  font-size: 14px;
  line-height: 12px;
  position: absolute;
  -webkit-transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  -webkit-transform: translate(0,24px);
  -ms-transform: translate(0,24px);
  transform: translate(0,24px);
  pointer-events: none;
}

.emotion-1._pebble_input_label_focused {
  font-size: 12px;
  line-height: 10px;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
  color: #6B7785;
}

.emotion-2 {
  height: 2px;
  background-color: #6161FF;
  margin-top: -2px;
  position: relative;
  width: 0;
  -webkit-transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  z-index: 9;
}

.emotion-2._pebble_input_highlight_state,
.emotion-2._pebble_input_highlight_focused {
  width: 100%;
}

.emotion-4 {
  position: absolute;
  top: 0;
  bottom: 0;
  margin: auto;
  height: 10px;
  right: 7px;
  color: #9FA6B0;
  font-size: 10px;
}

.emotion-4.__pebble__select__open {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

<div
  className="emotion-7"
>
  <div
    className="emotion-6"
  >
    <div
      style={
        Object {
          "display": "inline-block",
          "width": "100%",
        }
      }
    >
      <div
        className="emotion-5"
        onClick={[Function]}
      >
        <div
          className=" emotion-3"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <input
            aria-label="Choose Option"
            className="emotion-0"
            onChange={[Function]}
            readOnly={true}
            value=""
          />
          <label
            className=" emotion-1"
          >
            Choose Option
          </label>
          <div
            className="_pebble_input_highlight_read_only emotion-2"
            style={
              Object {
                "backgroundColor": "#6161FF",
              }
            }
          />
          
        </div>
        <i
          className="pi pi-arrow-drop-down emotion-4"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Component: Select single-select: snapshot 1`] = `
.emotion-7 {
  margin-bottom: 20px;
}

.emotion-6 {
  position: relative;
}

.emotion-5 {
  cursor: pointer;
  position: relative;
}

.emotion-3 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #ffffff;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 20px;
  height: 68px;
  pointer-events: none;
  margin-bottom: 0;
}

.emotion-3._pebble_input_wrapper_textarea {
  height: 110px;
}

.emotion-0 {
  outline: 0;
  border: 0;
  border-bottom: 1px solid #EDEDED;
  padding: 24px 0 12px 0;
  height: 48px;
  border-radius: 0;
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  width: 100%;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: #6B7785;
  color: #101721;
  padding-right: 15px;
}

.emotion-0:disabled {
  background-color: #ffffff;
}

.emotion-0::-webkit-input-placeholder {
  color: #E0E0E0;
}

.emotion-0::-moz-placeholder {
  color: #E0E0E0;
}

.emotion-0:-ms-input-placeholder {
  color: #E0E0E0;
}

.emotion-0:-moz-placeholder {
  color: #E0E0E0;
}

.emotion-0::-webkit-input-placeholder {
  color: #E0E0E0;
}

.emotion-0::-moz-placeholder {
  color: #E0E0E0;
}

.emotion-0:-ms-input-placeholder {
  color: #E0E0E0;
}

.emotion-0::placeholder {
  color: #E0E0E0;
}

.emotion-1 {
  color: #9FA6B0;
  font-size: 14px;
  line-height: 12px;
  position: absolute;
  -webkit-transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  -webkit-transform: translate(0,24px);
  -ms-transform: translate(0,24px);
  transform: translate(0,24px);
  pointer-events: none;
}

.emotion-1._pebble_input_label_focused {
  font-size: 12px;
  line-height: 10px;
  -webkit-transform: translate(0,0);
  -ms-transform: translate(0,0);
  transform: translate(0,0);
  color: #6B7785;
}

.emotion-2 {
  height: 2px;
  background-color: #6161FF;
  margin-top: -2px;
  position: relative;
  width: 0;
  -webkit-transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  transition: all 0.3s cubic-bezier(.64,.09,.08,1);
  z-index: 9;
}

.emotion-2._pebble_input_highlight_state,
.emotion-2._pebble_input_highlight_focused {
  width: 100%;
}

.emotion-4 {
  position: absolute;
  top: 0;
  bottom: 0;
  margin: auto;
  height: 10px;
  right: 7px;
  color: #9FA6B0;
  font-size: 10px;
}

.emotion-4.__pebble__select__open {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

<div
  className="emotion-7"
>
  <div
    className="emotion-6"
  >
    <div
      style={
        Object {
          "display": "inline-block",
          "width": "100%",
        }
      }
    >
      <div
        className="emotion-5"
        onClick={[Function]}
      >
        <div
          className=" emotion-3"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          <input
            aria-label="Choose Option"
            className="emotion-0"
            onChange={[Function]}
            readOnly={true}
            value=""
          />
          <label
            className=" emotion-1"
          >
            Choose Option
          </label>
          <div
            className="_pebble_input_highlight_read_only emotion-2"
            style={
              Object {
                "backgroundColor": "#6161FF",
              }
            }
          />
          
        </div>
        <i
          className="pi pi-arrow-drop-down emotion-4"
        />
      </div>
    </div>
  </div>
</div>
`;
